pkg/apis/secretgen2/v1alpha1/secret_template.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	sgv1alpha1 "github.com/vmware-tanzu/carvel-secretgen-controller/pkg/apis/secretgen/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name=Description,JSONPath=.status.friendlyDescription,description=Friendly description,type=string
+// +kubebuilder:printcolumn:name=Age,JSONPath=.metadata.creationTimestamp,description=Time since creation,type=date
+type SecretTemplate struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object metadata; More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec SecretTemplateSpec `json:"spec"`
+	// +optional
+	Status SecretTemplateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type SecretTemplateList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []SecretTemplate `json:"items"`
+}
+
+type SecretTemplateSpec struct {
+	//A list of input resources that are used to construct a new secret. Input Resources can refer to ANY Kubernetes API.
+	//If loading more than Secrets types ensure that `.spec.ServiceAccountName` is set to an appropriate value.
+	//Input resources are read in the order they are defined. An Input resource's name can be evaluated dynamically from data in a previously evaluated input resource.
+	InputResources []InputResource `json:"inputResources"`
+
+	//A JSONPath based template that can be used to create Secrets.
+	// +optional
+	JSONPathTemplate JSONPathTemplate `json:"template,omitempty"`
+
+	//The Service Account used to read inputResources. If not specified, then the Service Account of
+	//secretgen controller is used.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
```

Comment:
I guess you're optimizing for usability by making this optional but this smells like it has the potential for privilege escalation, in particular if this API will ever evolve to support input resources from multiple namespaces. But even for the single namespace case this could be problematic. What are the RBAC permissions of the service account secretgen controller uses today? What will they be in the future? If they are (or ever will be) for anything else than secrets in the same namespace as the template itself, there's a potential attack vector.

Reply:
Yeah this was for convenience as suggested in https://github.com/vmware-tanzu/carvel-secretgen-controller/issues/54 . To guard against this we can ensure that if a ServiceAccount is not specified then we only allow loading secrets in the same namespace. In the future if we add namespaces support we must ensure that any cross namespace loading of any resource requires a ServiceAccount.

Reply:
How do you feel about my concern w/r to the single namespace use case? Even if the SA for the secretgen controller is locked down to only `secrets` right now, how do we make sure it stays that way and we don't accidentally introduce the potential for privilege escalation as a side effect of an unrelated change elsewhere in this repo?

Reply:
If you want to get to the level of convenience you're after without coupling this to the RBAC of the controller SA, would it make sense to have a separate SA that is deployed alongside secretgen controller and used for the sole purpose to act as the default for `secrettemplates`? That SA could be named accordingly and we could bind it to a dedicated clusterrole that is **and forever will be** locked down to `secrets`.

Reply:
I'm suggesting we still use the secret-gen controller client but programmatically ensure we only ever load Secrets from the same namespace as SecretTemplate and error if any other type of inputresource/namespace is specified. That way we guard against your concern that secret-gen controller may gain permission in the future.